File: lib/crypto/crypto_exception.dart
/// Stable error codes shared with the Android (`CryptoException.kt`) and
/// iOS (`CryptoError.swift`) native implementations.
enum CryptoErrorCode {
  invalidFormat,
  keyNotFound,
  keyGenerationFailed,
  encryptionFailed,
  decryptionFailed,
  authenticationFailed,
  ioFailure,
  cancelled,
  invalidArgument,
  unknown;

  static CryptoErrorCode fromWire(String? code) {
    switch (code) {
      case 'INVALID_FORMAT':
        return CryptoErrorCode.invalidFormat;
      case 'KEY_NOT_FOUND':
        return CryptoErrorCode.keyNotFound;
      case 'KEY_GENERATION_FAILED':
        return CryptoErrorCode.keyGenerationFailed;
      case 'ENCRYPTION_FAILED':
        return CryptoErrorCode.encryptionFailed;
      case 'DECRYPTION_FAILED':
        return CryptoErrorCode.decryptionFailed;
      case 'AUTHENTICATION_FAILED':
        return CryptoErrorCode.authenticationFailed;
      case 'IO_FAILURE':
        return CryptoErrorCode.ioFailure;
      case 'CANCELLED':
        return CryptoErrorCode.cancelled;
      case 'INVALID_ARGUMENT':
        return CryptoErrorCode.invalidArgument;
      default:
        return CryptoErrorCode.unknown;
    }
  }
}

/// Thrown by every [SeroCrypto] operation on failure. Carries a typed
/// [code] so callers can branch on failure kind (e.g. show "wrong key" for
/// [CryptoErrorCode.authenticationFailed]) without parsing message strings.
class CryptoException implements Exception {
  final CryptoErrorCode code;
  final String message;

  const CryptoException(this.code, this.message);

  /// True when decryption failed because the GCM authentication tag did not
  /// verify - i.e. the wrong key was used, or the data was corrupted/tampered.
  bool get isAuthenticationFailure => code == CryptoErrorCode.authenticationFailed;

  /// True when the operation was cancelled via [SeroCrypto.cancel].
  bool get isCancelled => code == CryptoErrorCode.cancelled;

  @override
  String toString() => 'CryptoException(${code.name}): $message';
}

File: android/app/src/main/kotlin/com/bander/sero/crypto/CryptoException.kt
package com.bander.sero.crypto

/** Base type for every error the crypto engine can raise. Carries a stable
 * [code] so the Dart side can branch on failure kind without parsing strings. */
sealed class CryptoException(val code: String, message: String, cause: Throwable? = null) :
    Exception(message, cause) {

    class InvalidFormat(message: String) : CryptoException("INVALID_FORMAT", message)
    class KeyNotFound(alias: String) : CryptoException("KEY_NOT_FOUND", "No key found for alias '$alias'")
    class KeyGenerationFailed(message: String, cause: Throwable? = null) :
        CryptoException("KEY_GENERATION_FAILED", message, cause)
    class EncryptionFailed(message: String, cause: Throwable? = null) :
        CryptoException("ENCRYPTION_FAILED", message, cause)
    class DecryptionFailed(message: String, cause: Throwable? = null) :
        CryptoException("DECRYPTION_FAILED", message, cause)
    class AuthenticationFailed(message: String = "GCM tag verification failed: data is corrupted or tampered") :
        CryptoException("AUTHENTICATION_FAILED", message)
    class IoFailure(message: String, cause: Throwable? = null) :
        CryptoException("IO_FAILURE", message, cause)
    class Cancelled(message: String = "Operation was cancelled") : CryptoException("CANCELLED", message)
    class InvalidArgument(message: String) : CryptoException("INVALID_ARGUMENT", message)
}
